service/udp.go

```diff
@@ -254,16 +248,146 @@ func (h *packetHandler) validatePacket(textData []byte) ([]byte, *net.UDPAddr, *
 	return payload, tgtUDPAddr, nil
 }
 
+type AssociationHandleFunc func(ctx context.Context, conn net.Conn)
+
+// PacketServe listens for UDP packets on the provided [net.PacketConn] and creates
+// and manages NAT associations. It uses a NAT map to track active associations and
+// handles their lifecycle.
+func PacketServe(clientConn net.PacketConn, assocHandle AssociationHandleFunc, metrics NATMetrics) {
+	nm := newNATmap()
+	ctx, contextCancel := context.WithCancel(context.Background())
+	defer contextCancel()
+
+	for {
+		lazySlice := readBufPool.LazySlice()
+		buffer := lazySlice.Acquire()
+
+		expired := false
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					slog.Error("Panic in UDP loop. Continuing to listen.", "err", r)
+					debug.PrintStack()
+					lazySlice.Release()
+				}
+			}()
+			n, addr, err := clientConn.ReadFrom(buffer)
+			if err != nil {
+				lazySlice.Release()
+				if errors.Is(err, net.ErrClosed) {
+					expired = true
+					return
+				}
+				slog.Warn("Failed to read from client. Continuing to listen.", "err", err)
+				return
+			}
+			pkt := &packet{payload: buffer[:n], done: lazySlice.Release}
+
+			// TODO(#19): Include server address in the NAT key as well.
+			assoc := nm.Get(addr.String())
+			if assoc == nil {
+				assoc = &association{
+					pc:     clientConn,
+					raddr:  addr,
+					readCh: make(chan *packet, 5),
+				}
+				if err != nil {
+					slog.Error("Failed to handle association", slog.Any("err", err))
+					return
+				}
+
+				var existing bool
+				assoc, existing = nm.Add(addr.String(), assoc)
+				if !existing {
+					metrics.AddNATEntry()
+					go func() {
+						assocHandle(ctx, assoc)
+						metrics.RemoveNATEntry()
+						nm.Del(addr.String())
+					}()
+				}
+			}
+			select {
+			case assoc.readCh <- pkt:
```

Comment:
You must not write to a closed channel. You need a way to ensure that.

Perhaps you need an assoc.doneCh that you check here. You can close doneCh when the association is done.
The select could test against doneCh here, and call nm.Del here. This way there's no race condition.

You'd be effectively sending an "event" to be processed in the single-thread "event loop".